src/besselj.jl

```diff
@@ -150,3 +150,222 @@ function besselj1(x::Float32)
+
+        αp = evalpoly((1/x)^2, (s0, s1, s2, s3, s4, s5, s6, s7, s8, s9, s10, s11, s12, s13, s14, s15, s16, s17, s18, s19, s20, s21, s22, s23, s24, s25, s26, s27, s28))
+        α = x * (evalpoly((1/x)^2, (s0, -s1, -s2/3, -s3/5, -s4/7, -s5/9, -s6/11, -s7/13, -s8/15, -s9/17, -s10/19, -s11/21, -s12/23, -s13/25, -s14/27, -s15/29, -s16/31, -s17/33, -s18/35, -s19/37, -s20/39, -s21/41, -s22/43, -s23/45, -s24/47, -s25/49, -s26/51, -s27/53, -s28/55)))
+    else
+        αp = evalpoly((1/x)^2, (s0, s1, s2, s3, s4, s5, s6))
+        α = x * (evalpoly((1/x)^2, (s0, -s1, -s2/3, -s3/5, -s4/7, -s5/9, -s6/11)))
+    end
+    α = α - T(pi)/4 - T(pi)/2 * v
+    b = sqrt(2 / T(pi)) / sqrt(αp * x)
+
+    return α, b
+end
+# to acurrately use x > 1.7*nu
+function besselj_large_argument(v, x::T) where T
+    α, b = _α_αp_asymptotic(v, x)
+    return cos(α)*b
+end
+function bessely_large_argument(v, x::T) where T
+    α, b = _α_αp_asymptotic(v, x)
+    return sin(α)*b
+end
+
+# this needs a better way to sum these as it produces large errors
+# only valid in non-oscillatory regime (v>1/2, 0<t<sqrt(v^2 - 0.25))
+# power series has premature underflow for large orders
+# gives wrong answers for x > 20.0 (might want to fix this)
+function besselj_small_arguments_orders(v, x::T) where T
+    MaxIter = 1000
+    out = zero(T)
+    a = (x/2)^v / gamma(v + one(T))
+    t2 = (x/2)^2
+    for i in 0:MaxIter
+        out += a
+        abs(a) < eps(T) * abs(out) && break
+        a *= -inv((v + i + one(T)) * (i + one(T))) * t2
+    end
+    return out
+end
+
+# this needs a better way to sum these as it produces large errors
+# use when v is large and x is small
+# need for bessely 
+function log_besselj_small_arguments_orders(v, x::T) where T
```

Comment:
We need a better way to calculate these power series. There is a lot of error propagation in the loops because the first 10 terms or so actually grow and oscillate...
```julia
julia> Bessels.log_besselj_small_arguments_orders(big"7", big"17.1")
0.1794205811198151334600691013458264941056284393401507757350616668754102873422111

julia> Bessels.log_besselj_small_arguments_orders(7, 17.1)
0.17942058112702328
```
Is there perhaps a better way to compensate for this?

Reply:
Is there any possible simplification from unrolling the first couple terms?

Reply:
Ya I'm thinking because I have such tight bounds on when this is used `x<20` it might be worth just completely unrolling this and just using evalpoly for the max number of terms..? I think that would definitely be more accurate and probably faster than checking for convergence at each iteration?

Reply:
yeah that will be a lot faster since you can then make it only require 1 division rather than 1 division per step.

Reply:
Should probably set this up better because I am still losing a lot of precision.
```julia
function t(v, x)
x2 = (x/2)^2

b = muladd(34, v, 1156)
a = muladd(-x2, inv(b), 1)

b = x2 / muladd(33, v, 1089)
a = muladd(-a, b, 1)
b = x2 / muladd(32, v, 1024)
a = muladd(-a, b, 1)
b = x2 / muladd(31, v, 961)
a = muladd(-a, b, 1)
b = x2 / muladd(30, v, 900)
a = muladd(-a, b, 1)
b = x2 / muladd(29, v, 841)
a = muladd(-a, b, 1)
b = x2 / muladd(28, v, 784)
a = muladd(-a, b, 1)
b = x2 / muladd(27, v, 729)
a = muladd(-a, b, 1)
b = x2 / muladd(26, v, 676)
a = muladd(-a, b, 1)
b = x2 / muladd(25, v, 625)
a = muladd(-a, b, 1)
b = x2 / muladd(24, v, 576)
a = muladd(-a, b, 1)
b = x2 / muladd(23, v, 529)
a = muladd(-a, b, 1)
b = x2 / muladd(22, v, 484)
a = muladd(-a, b, 1)
b = x2 / muladd(21, v, 441)
a = muladd(-a, b, 1)
b = x2 / muladd(20, v, 400)
a = muladd(-a, b, 1)
b = x2 / muladd(19, v, 361)
a = muladd(-a, b, 1)
b = x2 / muladd(18, v, 324)
a = muladd(-a, b, 1)
b = x2 / muladd(17, v, 289)
a = muladd(-a, b, 1)
b = x2 / muladd(16, v, 256)
a = muladd(-a, b, 1)
b = x2 / muladd(15, v, 225)
a = muladd(-a, b, 1)
b = x2 / muladd(14, v, 196)
a = muladd(-a, b, 1)
b = x2 / muladd(13, v, 169)
a = muladd(-a, b, 1)
b = x2 / muladd(12, v, 144)
a = muladd(-a, b, 1)
b = x2 / muladd(11, v, 121)
a = muladd(-a, b, 1)
b = x2 / muladd(10, v, 100)
a = muladd(-a, b, 1)
b = x2 / muladd(9, v, 81)
a = muladd(-a, b, 1)
b = x2 / muladd(8, v, 64)
a = muladd(-a, b, 1)
b = x2 / muladd(7, v, 49)
a = muladd(-a, b, 1)
b = x2 / muladd(6, v, 36)
a = muladd(-a, b, 1)
b = x2 / muladd(5, v, 25)
a = muladd(-a, b, 1)
b = x2 / muladd(4, v, 16)
a = muladd(-a, b, 1)
b = x2 / muladd(3, v, 9)
a = muladd(-a, b, 1)
b = x2 / muladd(2, v, 4)
a = muladd(-a, b, 1)
b = x2 / muladd(1, v, 1)
a = muladd(-a, b, 1)
return a*(x/2)^v * 1 / (SpecialFunctions.gamma(v+1))
end

julia> v = 7
7

julia> x = 17.1
17.1

julia> 1 - SpecialFunctions.besselj(v, x) / t(v,x)
4.896916205865409e-11

julia> v = big"7"
7

julia> x = big"17.1"
17.10000000000000000000000000000000000000000000000000000000000000000000000000006

julia> 1 - SpecialFunctions.besselj(v, x) / t(v,x)
2.11919566953799423385499459990264416821865652645594220138087841875073985051455e-19
```

Reply:
if you can find a maximum number of arguments, there is a major speedup here. You can multiply the whole equation by `(n+1)(n+2)(n+3)...` before hand (and you can fairly easily compensate this multiplication), and then you only have 1 division to do.

Reply:
For `x<10` we need 25 terms and for `x<20` we need 40 terms. 

That was my current cutoff for this method `x<20` as for larger x it was harder to get accurate.

Reply:
you can do way better than that if you have bounds on the order, right?

Reply:
Most definitely. But it doesn't equally scale for very large orders. We need about half the terms for orders above 60 which decrease slightly for increasing order.

For `x=20.0, nu = 0.5` we need 40 terms.
For `x = 20.0, nu = 60.5` we need 20 terms.

Reply:
if my math is correct, for `v>45` and `x<20`, 10 terms should suffice.

Reply:
that seems too pesimistic to me. Isn't the condition `x^2*gamma(v)/gamma(v+k)<2^-52`?

Reply:
also, we can deal with the cases where `x>>nu` using `besselj_large_argument`

Reply:
That looks close to me quickly looking... I got those numbers from just outputting the amount of each interations from the loop it took to converge. 
```julia

julia> x=20.0; v = 0.5; k = 40
40

julia> x^2*gamma(v)/gamma(v+k)
5.51286033593326e-45
```

Reply:
And for reference. The hypergeometric functions don't provide better accuracy..
```julia
using HypergeometricFunctions, SpecialFunctions
function besselj_hyper(v, x)
       a = HypergeometricFunctions.drummond0F1(v + 1, -x^2/4)
       return (x/2)^v / gamma(v+1) * a
end

julia> besselj_hyper(7.0, 17.1)
0.17915365638067893

julia> Float64(besselj_hyper(big"7.0", big"17.1"))
0.1791536563807948

julia> Float64(SpecialFunctions.besselj(big"7.0", big"17.1"))
0.1791536563807948
```